### PR TITLE
(2 of 2) of the rest of the new bytecode emitter

### PR DESCRIPTION
**Pull request by magarciaEPFL**
This PR completes the new bytecode emitter, including building and writing class files in parallel, by building upon https://github.com/scala/scala/pull/2863

Before writing class files, a modicum of code tightening transformations are necessary: unreachable code elimination, which in turn calls for cleanup of labels and exception entries made redundant by unreachable code elimination.

In this PR, each such transformation is introduced in a commit with tests for the transformation in isolation. Another commit also in this PR tests each such transformation in connection with the bytecode emitter.

I'll suggest a reviewer once tests pass, comments and reviews welcome.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/jvm/GenBCode.scala [hunk 1]
```diff
@@ -58,6 +111,15 @@ abstract class GenBCode extends BCodeSyncAndTry {
     override def description = "Generate bytecode from ASTs using the ASM library"
     override def erasedTypes = true
 
+    // number of woker threads for pipeline-2 (the pipeline in charge of most optimizations except inlining).
+    val MAX_THREADS = scala.math.min(
+      4,
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/GenBCode.scala]: We need a way to configure this. I'd suggest a -Y compiler option. Use case would be to disable parallelism altogether here to work around a bug, or when investigating a bug.

A system property is an alternative, but that is JVM-wide and unsuitable for platforms that run multiple compilers in one JVM (e.g. SBT).

File: src/compiler/scala/tools/nsc/backend/bcode/DanglingExcHandlers.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ *  @version 1.0
+ *
+ */
+public class DanglingExcHandlers {
+
+    /** after transform() has run, this field records whether
+     *  at least one pass of this transformer modified something. */
+    public boolean changed = false;
+
+    public void transform(final MethodNode mn) {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/bcode/DanglingExcHandlers.java:51]: Exposing a naked field `changed` like this is sort of frowned upon by Javans, no?

You could encapsulate it and capture the similarity between these classes with:

```
interface Transformer {
  boolean changed();
  void transform(MethodNode mn)
}
```

**Comment by magarciaEPFL** [comment]
Thanks for the comments so far. I'll give more time for others to chime in, to then add:
- a single commit with all documentation and readability improvements (eg removing the `_root_` prefixes)
- one commit per behavioral change (eg adding `-Y` option to choose the number of worker threads, from 1 to N, etc.)

**Comment by magarciaEPFL** [comment]
Commits eb48a8c5e0a160296aff696d1f292141ea9952c0 and bdafe6dc9f65db1bfc827f147ef93a1eb60a3b12 incorporate review feedback. However, another commit still pending, to add a `-Y` compiler option to configure the number of worker threads. Coming soon.

File: src/compiler/scala/tools/nsc/backend/jvm/GenBCode.scala (hunk 1 above)
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/backend/jvm/GenBCode.scala]: I agree.

Would it also make sense to make ExecutionContext pluggable so people can't control thread pools when running Scala compiler in an embedded context?

**Comment by gkossakowski** [comment]
@magarciaEPFL: this PR looks really good. Thanks for taking time and writing proper unit tests for your ASM transformations! Could you rewrite all of them to be JUnit tests, though?

In 54a7f806f0cb50e386d2800756f9737d720ef732 you do test bytecode generated by Scala compiler so you can leave that test as-is.

**Comment by magarciaEPFL** [comment]
@gkossakowski , @retronym A few more commits addressing the pending items:
- converting tests to JUnit 
- compiler option for number of threads

are being worked on, and should be ready in a day or two.

It would be great if this PR were merged as-is so that early adopters can field-test GenBCode with M5 and provide us with much needed feedback. Quite likely the pending improvements will also be ready on time, but this PR is the main thing.

Rest assured that the next commits I'll include (in that case, in a follow-up PR) will address the pending items.

**Comment by magarciaEPFL** [comment]
One of the comments in a previous PR emphasized a distinction between integration and unit tests in the context of ASM transformations. This PR as originally submitted contains both kinds of tests.

The commits just added refactor the units tests into JUnit tests, as requested. The actual work performed by old and new unit tests hasn't changed.

Now that both integration and junit tests have been addressed, it only remains to add a compiler option to configure the number of worker threads in GenBCode.

As already mentioned, this PR is correct, and can be merged. The compiler option in question will be added soon.

**Comment by gkossakowski** [comment]
@magarciaEPFL: we'll need a little bit more polish when it comes to unit tests. I'll try to work on that tonight to demonstrate what I mean by that.

**Comment by magarciaEPFL** [comment]
@gkossakowski , I don't know what kind of polish in unit tests it's so important to show off, all I'm saying is this PR makes the compiler faster. Would have made, had it been merged in M5.

**Comment by gkossakowski** [comment]
@magarciaEPFL: it's not about "showing off" but merging code that is of high quality. All I'm offering is a little guidance on how to write good unit tests.

Also, I'm very much aware of benefits of the new backend. However, that's not what we are discussing it here.

**Comment by gkossakowski** [comment]
@magarciaEPFL: I reworked your tests to become idiomatic JUnit tests.

Here's me showing off:
Me showing off about my tests

More seriously, see gkossakowski@3c80c57a44745789f6ce46981319f26edce223ed and @gkossakowski@bcf58bd2fb71741e6c220a3ab3f7b5edffcd3603. Please cherry pick the first commit and squash the second one into your 1aacd32. Once done, push the final result this this PR. I'll review everything today. If there will be no problems I'll merge it today and your PR will make it into M5.

**Comment by magarciaEPFL** [comment]
@gkossakowski 

I don't know what I like most,
- the animated GIF
- seeing a few lines shorter `ant test.junit` output
- or having to clean `build/junit` before running the updated tests

Anyway, it's late here so all I've managed to do is:
- branch from the last commit that remains as is ( https://github.com/magarciaEPFL/scala/commit/bdafe6dc9f65db1bfc827f147ef93a1eb60a3b12 )
- cherry-pick your commit to upgrade to JUnit 4.11
- squash the not-quite-idiomatic and the idiomatic-show-off commits
- add the commit to configure the number of threads

The result can be found in branch https://github.com/magarciaEPFL/scala/tree/backendish46 that I'm submitting as PR https://github.com/scala/scala/pull/2908

**Comment by magarciaEPFL** [comment]
OK, thanks to all reviewers and I'll find out tomorrow what M5 is good for :smile:

**Comment by som-snytt** [comment]
I have a commit in a PR somewhere to fix:

> having to clean build/junit before running the updated tests

which is totally insanity-inducing.

But do you wish you had to ant partest.task if you happen to touch code under partest-extras?

Monolithicity has its privileges.

But it's all in the right direction. NNE.

**Comment by gkossakowski** [comment]
@magarciaEPFL: In the future please keep amending existing PR instead of opening a new one. It makes it easier to manage things (we have 30 PRs open at the moment) for us and we don't lose the context like existing comments.

All newly opened PRs are automatically assigned to M6 release. I'll reassign your #2908 to M5 manually but please keep that in mind in the future.

**Pull request closed without merging**